odoo/tools/date_utils.py

```diff
@@ -59,6 +82,118 @@ def to_timezone(tz: tzinfo | None) -> Callable[[datetime], datetime]:
     return lambda dt: dt.astimezone(tz)
 
 
+def parse_iso_date(value: str) -> date | datetime:
+    """ Parse a ISO encoded string to a date or datetime.
+
+    :raises ValueError: when the format is invalid or has a timezone
+    """
+    # Looks like ISO format
+    if len(value) <= 10:
+        return date.fromisoformat(value)
+    now = datetime.fromisoformat(value)
+    if now.tzinfo is not None:
+        raise ValueError(f"expecting only datetimes with no timezone: {value!r}")
+    return now
+
+
+def parse_date(value: str, env: Environment) -> date | datetime:
+    r""" Parse a technical date string into a date or datetime.
+
+    This supports ISO formatted dates and dates relative to now.
+    `parse_iso_date` is used if the input starts with r'\d+-'.
+    Otherwise, the date is computed by starting from now at user's timezone.
+    We can also start 'today' (resulting in a date type). Then we apply offsets:
+
+    - we can add 'd', 'w', 'm', 'y', 'H', 'M', 'S':
+      days, weeks, months, years, hours, minutes, seconds
+      - "+3d" to add 3 days
+      - "-1m" to subtract one month
+    - we can set a part of the date which will reset to midnight or only lower
+      date parts
+      - "=1d" sets first day of month at midnight
+      - "=6m" sets June and resets to midnight
+      - "=3H" sets time to 3:00:00
+    - weekdays are handled similarly
+      - "=tuesday" sets to Tuesday of the current week (starting Monday) at midnight
+      - "+monday" goes to next Monday (no change if we are on Monday)
+
+    The DSL for relative dates is as follows:
+    ```
+    relative_date := ('today' | 'now')? offset*
+    offset := date_rel | time_rel | weekday
+    date_rel := (regex) [=+-]\d+[dwmy]
+    time_rel := (regex) [=+-]\d+[HMS]
+    weekday := [=+-] ('monday' | ... | 'sunday')
+    ```
+
+    :param value: The string to parse
+    :param env: The environment to get the current date (in user's tz)
+    :param naive: Whether to cast the result to a naive datetime.
+    """
+    if re.match(r'\d+-', value):
+        return parse_iso_date(value)
+    terms = value.split()
+    if not terms:
+        raise ValueError("Empty date value")
+
+    # Find the starting point
+    from odoo.orm.fields_temporal import Date, Datetime  # noqa: PLC0415
+    now: datetime | date = Datetime.now()
```

Comment:
This variable contains the result of operations.  It is not meant to represent now or today, hence the name `now` is a bit confusing.  Why don't you call it `result` or `dt` instead?